module/builder/collection/builder.go

```diff
@@ -60,277 +53,244 @@ func NewBuilder(
 		apply(&b.config)
 	}
 
-	return &b
+	// sanity check config
+	if b.config.ExpiryBuffer >= flow.DefaultTransactionExpiry {
+		return nil, fmt.Errorf("invalid configured expiry buffer exceeds tx expiry (%d > %d)", b.config.ExpiryBuffer, flow.DefaultTransactionExpiry)
+	}
+
+	return &b, nil
 }
 
 // BuildOn creates a new block built on the given parent. It produces a payload
 // that is valid with respect to the un-finalized chain it extends.
 func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) error) (*flow.Header, error) {
-	var proposal cluster.Block
+	var proposal cluster.Block                 // proposal we are building
+	var parent flow.Header                     // parent of the proposal we are building
+	var clusterChainFinalizedBlock flow.Header // finalized block on the cluster chain
+	var refChainFinalizedHeight uint64         // finalized height on reference chain
+	var refChainFinalizedID flow.Identifier    // finalized block ID on reference chain
 
 	startTime := time.Now()
 
+	// STEP ONE: Load some things we need to do our work.
```

Comment:
```suggestion
	// STEP ONE: build a lookup for excluding duplicated transactions. This is briefly how it works:
	// expired transaction won't be included in new collection block, so it's not possible for expired tx
	// to be duplicated.
	// therefore, we only need a lookup for all non-expired transactions.
	// non-expired transactions only exists in non-expired collections, because a valid collection
	// must only include non-expired transactions.
	// in other words, we can build the transaction lookup by finding all the non-expired collections.
	// Note, in order to make the search algorithm simpler, finding all non-expired collections might 
	// end up including some expired collections. But that's OK, because expired transactions are
	// excluded in new collection block anyway.
	// a non-expired collection can be finalized or un-finalized. 
	// - to find all non-expired and finalized collections, we make use of an index 
	//    (main_chain_finalized_height -> cluster_block_id), to search for a range of main chain heights 
	//     which could be only referenced by non-expired collections. 
	// - to find all non-expired and un-finalized collections, we can't use the above index, because it's
	//    only for finalized collections. Instead, we simply traverse along the chain up to the last 
	//    finalized block. This could possibly include some expired collections, but it's OK as explained
	//    above.
	// after combining both the finalized and un-finalized collection blocks that are not expired, 
	// we can iterate through their transactions, and build a lookup for excluding duplicated transactions.
```

Reply:
Thanks for this - I adjusted the wording a bit to be more specific: we are searching for collections which have an overlapping expiry window with our collection under construction. Not necessarily collections which are expired vs not expired.